### PR TITLE
[java-gen] Adding support for default values (Serialization::unmarshal based)

### PR DESCRIPTION
**Pull request by fabiobrz**
## Description

Draft/attempt to support default values generation in a different way than in https://github.com/fabric8io/kubernetes-client/pull/3944. 

The approach here is to delegate the actual deserialization of the default value (as a `JsonNode` content) to the `Serialization::unmarshal` method, which is used already extesilvely in the client.

Consequently, the generated code will depend on `io.fabric8.kubernetes.client` classes and the actual deserialization will take place at runtime.

Fixes https://github.com/fabric8io/kubernetes-client/issues/3869

FYI @andreaTP 

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java [hunk 1]
```diff
@@ -212,11 +215,20 @@ public GeneratorResult generateJava() {
         if (!prop.isNullable) {
           // from https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting-and-nullable :
           // "_null values for fields that either don't specify the nullable flag, or give it a false
-          // value, will be pruned before defaulting happens..._"
+          // value, will be pruned before defaulting happens. If a default is present, it will be applied_"
           objField.addAnnotation(
               new SingleMemberAnnotationExpr(
                   new Name("com.fasterxml.jackson.annotation.JsonSetter"),
                   new NameExpr("nulls = com.fasterxml.jackson.annotation.Nulls.SKIP")));
+          if (prop.defaultValue != null) {
```
Comment (andreaTP) [inline java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java:223]: Have you attempted to move this block out from the `if` for the nullable?
If we have set all the Jackson annotations properly this might work according to spec even defining always the body block.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java [hunk 2]
```diff
@@ -212,11 +215,20 @@ public GeneratorResult generateJava() {
...
+            // always add an "init_*" method which will be expanded by statements needed to create the default value
```
Comment (andreaTP) [inline java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java:224]: The actual body of the method is trivial, I would suggest to avoid generating additional methods that can possibly cause collisions in edge cases. Can you directly assign the block expression generated by `generateDefaultInitializerBody` ?

File: java-generator/core/src/test/java/io/fabric8/java/generator/GeneratorTest.java [hunk 3]
```diff
@@ -504,4 +511,60 @@ void testClassNamesDisambiguationWithPackageNesting() {
     assertTrue(clzO2.get().getFieldByName("o2").isPresent());
     assertTrue(clzO2.get().getFieldByName("o3").isPresent());
   }
+
+  @Test
+  void testObjectWithDefaultFields() {
```
Comment (andreaTP) [inline java-generator/core/src/test/java/io/fabric8/java/generator/GeneratorTest.java:516]: this test looks pretty complicated, as we already have compilation and approval tests to verify the actual "content" of the default, I would be happy to only test that a default value exists/doesn't exists when is defined/when is not.

I would be actually more interested into using the integration tests with the Maven plugin to verify that the behaviour is according to specification e.g.: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting-and-nullable

**Review by andreaTP — CHANGES_REQUESTED** [review]
Thanks a lot @fabiobrz !

This is definitely going in the right direction!
The implementation is simple and straightforward (and can be slightly simplified even more 🙂  )

I added a few comments but I really like this PR already, thanks a lot!

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java [hunk 4]
```diff
@@ -54,10 +60,31 @@ public String getDescription() {
...
+  }
+
+  /**
+   * This method is responsible for creating a block statement that
+   * represents the body of a default value initialization method.
+   *
+   * @return a {@link com.github.javaparser.ast.stmt.BlockStmt} instance that contains all the statements that code the
+   *         default value initialization.
+   */
+  protected BlockStmt generateDefaultInitializerBody() {
```
Comment (andreaTP) [inline java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java:74]: I understand that this request might sounds a little "odd", but I would prefer if we can keep this encoding in the "parent" class (e.g. probably is sufficient a `private` method in JObject) and expose the `defaultValue` using a Getter.

The reason is:
currently the recursion is leveraging a single code "brach" that is implemented using `generateJava`, keeping it uniform makes it easier to refactor in the future and improve performance on the hot path.

Do you see any problem in implementing it in this way?

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java [hunk 5]
```diff
@@ -54,10 +60,31 @@ public String getDescription() {
...
+    BlockStmt body = new BlockStmt();
+    Expression unmarshallExpression = new NameExpr(
+        "io.fabric8.kubernetes.client.utils.Serialization.unmarshal(\"" + this.defaultValue.toString().replace("\"", "\\\"")
```
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java]: Fixed, let me know whether this fits in with what you meant.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java (hunk 1 above)
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java]: Right, fixed.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java (hunk 2 above)
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java]: Indeed, fixed.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java]: Yep, done as per your suggestion.

File: java-generator/core/src/test/java/io/fabric8/java/generator/GeneratorTest.java (hunk 3 above)
Comment (fabiobrz) [reply java-generator/core/src/test/java/io/fabric8/java/generator/GeneratorTest.java]: Test is now way more simple. Integration testing can be done in a different PR.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java (hunk 4 above)
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java]: I do agree, fixed.

**Comment by manusa** [comment]
Hi @andreaTP, could you please review again :pray:

**Comment by andreaTP** [comment]
Thanks again @fabiobrz , this implementation is really excellent! (especially enjoyed the use of `StringEscapeUtils`!)

I approve this since it's good to go, it would be nice if we can add (even in a follow up PR) an integration tests (maybe based on https://github.com/fabric8io/kubernetes-client/pull/3918 ) covering the actual behavior and interaction with nullable (e.g. taken from the kubernetes docs).

**Review by andreaTP — APPROVED** [review]

**Comment by fabiobrz** [comment]
> Thanks again @fabiobrz , this implementation is really excellent! (especially enjoyed the use of `StringEscapeUtils`!)

Thanks you as well, for providing valuable feedback n the changes.
 
> I approve this since it's good to go, it would be nice if we can add (even in a follow up PR) an integration tests (maybe based on #3918 ) covering the actual behavior and interaction with nullable (e.g. taken from the kubernetes docs).

Yes, since this belongs to a preview implementation I think that integration tests combining defaulting and nullable should be provided in a follow up PR. I'll log the issue, and maybe take care of it, since that's something I am interested in as well.

**Comment by fabiobrz** [comment]
Tiny code format change to trigger CI checks again.

**Review by manusa — APPROVED** [review]

**Review by oscerd — APPROVED** [review]

**Comment by andreaTP** [comment]
Nice! 🎉  Thanks!
